flaky tests: consistent logging

a-palchikov opened the pull request:
This PR simplifies the logger set up for tests: the logger is setup consistently with `DebugLevel` and outputting to `os.Stderr`, and formatter with timestamps since timing skews might be highly relevant for troubleshooting time-sensitive code.
I'm also bumping logrus to v1.6.0 since I found that our fork has data races and the fork is pretty far behind that release.
Running at `DebugLevel` also discovered a legitimate bug in roles code.

Requires https://github.com/gravitational/teleport.e/pull/183.

a-palchikov committed:
* Update logrus package to fix data races
 * Introduce a logger that uses the test context to log the messages so
   they are output if a test fails for improved trouble-shooting.

a-palchikov committed:
Revert introduction of test logger - simply leave logger configuration at debug level outputting to stderr during tests.

a-palchikov committed:
Run integration test for e as well

a-palchikov commented on lib/events/auditlog.go:
```diff
@@ -109,7 +109,7 @@ func init() {
 type AuditLog struct {
 	sync.Mutex
-	*log.Entry
+	log.FieldLogger
```
`logrus.Entry` in v1.6.0 embeds a `context.Context` which causes disambiguation errors here and below.

awly replied:
Could you refactor this to not embed the logger (name this field `log`)?
Otherwise we pollute the type's API for a few saved characters.

a-palchikov commented on lib/services/role.go:
```diff
@@ -1501,7 +1501,7 @@ func MatchLabels(selector Labels, target map[string]string) (bool, string, error
 // RoleNames returns a slice with role names. Removes runtime roles like
 // the default implicit role.
 func (set RoleSet) RoleNames() []string {
-	out := make([]string, len(set)-1)
```
this was misbehaving for `len(RoleSet) <= 1`

a-palchikov committed:
Use make with a cap and append to only copy the relevant roles.

awly commented on go.mod:
```diff
@@ -106,5 +105,4 @@ require (
 replace (
 	github.com/coreos/go-oidc => github.com/gravitational/go-oidc v0.0.3
 	github.com/iovisor/gobpf => github.com/gravitational/gobpf v0.0.1
-	github.com/sirupsen/logrus => github.com/gravitational/logrus v0.10.1-0.20171120195323-8ab1e1b91d5f
```
Do you know why we made a fork?
Hope we're not losing some internal patch that teleport depends on.

awly replied:
Awesome, hopefully we can archive our fork after this merges!

r0mant replied:
@a-palchikov With this change will `WithError(err)` finally start behaving properly by printing the error with a stack trace, or would we need to additionally bump logrus to get it working?

awly commented on lib/kube/proxy/forwarder.go:
```diff
@@ -216,7 +216,7 @@ func NewForwarder(cfg ForwarderConfig) (*Forwarder, error) {
 type Forwarder struct {
 	sync.Mutex
-	*log.Entry
+	log.FieldLogger
```
same here, please name the field and refactor all call sites.

awly commented on lib/reversetunnel/srv.go:
```diff
@@ -104,7 +104,7 @@ type server struct {
 	ctx context.Context
 
-	*log.Entry
+	log.FieldLogger
```
same here - name the field

awly commented on lib/utils/cli_test.go:
```diff
@@ -36,31 +36,34 @@ var _ = check.Suite(&CLISuite{})
 
 func (s *CLISuite) TestUserMessageFromError(c *check.C) {
+	// FIXME: enable after https://drone.gravitational.io/gravitational/teleport/3517 is merged
+	c.Skip("TODO")
```
Leaving a comment to resolve after this is fixed

awly reviewed (commented):
Great cleanup, thanks!

a-palchikov committed:
Address review comments

awly commented on lib/reversetunnel/srv_test.go:
```diff
@@ -38,7 +39,7 @@ func TestServerKeyAuth(t *testing.T) {
 	require.NoError(t, err)
 
 	s := &server{
-		Entry: testlog.FailureOnly(t),
+		log: utils.NewLoggerForTests(logrus.New()),
```
`testlog.FailureOnly` has a different behavior - it buffers all logs and only prints them when `t` fails.
please revert this change

a-palchikov replied:
I believe testing has this behavior by default - it buffers all output (sent via `t.Log(f)` or  `os.StdXXX`) and only outputs it if the test fails regardless of the `-v` (verbose) option. The verbose option is only relevant if one wants to see logs regardless of the outcome.
Do you mean that this should suppress log output regardless of verbosity and only output it on failure?

I initially implemented `InitLoggerForTests` to accept the value of `*testing.T` and route the logs using its `Log(f)` APIs but it became a bit of concern to use with a global logrus logger since it should not outlive a test so I abandoned this approach in favor of writing to `os.Stderr` as this is also buffered by the testing package with the behavior described above.

awly replied:
the default behavior in `testing` doesn't know which test outputs which logs.
so if you have a package with 100 tests and 1 of them fails, you get logs for all 100 tests intermixed with test result output.

and yes, the reason I wrote `testlog.FailureOnly` is because it's scoped to a single `testing.T` and not globally

a-palchikov replied:
The whole point of changes in this PR is exactly this behavior - hence I made this change too. Like you said, it is not going to be helpful if all of the tests' logs are output upon failure (which will happen if this is merged), so I might need to reconsider the approach I've taken with this.

The goal is getting the logs only for the failed test - no doubt about that but also that it should be universal - i.e. all tests should use the FailureOnly logger otherwise it's a hit or miss if used selectively. I'll think about a way to get this goal instead.

awly replied:
One other option is to set `logrus.Logger.Out` to a wrapper around `t.Log/c.Log`
It works for both `gocheck` and `testing` tests.

The challenge is making sure that none of our code uses the global logrus logger, but that can be done incrementally afterwards.

a-palchikov committed:
Update integration test suite to use test-local logger that would only
output logs iff a specific test has failed - no logs from other test
cases will be output.

a-palchikov committed:
Revert changes to InitLoggerForTests API

a-palchikov committed:
Create a new logger instance when applying defaults or merging with file service configuration

a-palchikov committed:
Introduce a local logger interface to be able to test file configuration
merge.

a-palchikov committed:
Fix kube integration tests w.r.t log

awly commented on lib/reversetunnel/srv_test.go:
```diff
@@ -38,7 +38,7 @@ func TestServerKeyAuth(t *testing.T) {
 
 	s := &server{
-		Entry: testlog.FailureOnly(t),
+		log: utils.NewLoggerForTests(),
```
let's revert this change, `testlog.FailureOnly` seems appropriate for non-gocheck tests

a-palchikov replied:
Yes, sorry - missed the spot.

awly commented on lib/service/service.go:
```diff
@@ -83,9 +84,7 @@ import (
-var log = logrus.WithFields(logrus.Fields{
-	trace.Component: teleport.ComponentProcess,
-})
+var log = logrus.WithField(trace.Component, teleport.ComponentProcess)
```
looks like there are still a few references to this package-global logger instead of `process.log` below.
remove this var and fix any compilation errors below.

awly commented on lib/srv/app/server.go:
```diff
@@ -440,6 +441,7 @@ func (s *Server) newHTTPServer() *http.Server {
 	return &http.Server{
 		Handler:           authMiddleware,
 		ReadHeaderTimeout: defaults.DefaultDialTimeout,
+		ErrorLog:          stdlog.New(utils.NewStdlogAdaptor(s.log.Error), teleport.ComponentApp, stdlog.LstdFlags),
```
this looks like a common pattern.
how about pulling it into the utils package, something like
```
func NewStdlog(logger leveledOutputFunc, component string) *stdlog.Logger
```

awly commented on lib/utils/cli.go:
```diff
@@ -207,6 +227,41 @@ func EscapeControl(s string) string {
 	return s
 }
 
+// Write writes the specifies buffer p to the underlying logger at INFO level.
```
```suggestion
// Write writes the specifies buffer p to the underlying logger.
```

awly commented on lib/utils/testlog/log.go:
```diff
@@ -20,19 +20,19 @@ package testlog
 // test fails.
-func FailureOnly(t *testing.T) *logrus.Entry {
-	// Collect all the output in buf.
+func FailureOnly(t testingInterface) utils.Logger {
```
the argument type should be exported

awly commented on lib/utils/testlog/log.go:
```diff
@@ -42,5 +42,64 @@ func FailureOnly(t *testing.T) *logrus.Entry {
+}
+
+// Cleanup registers the specified handler f to be run
+// after the hest gas completed
```
```suggestion
// after the test has completed
```

a-palchikov replied:
Addressed in https://github.com/gravitational/teleport/pull/4849/commits/756c17b66db83fc5637415d0231fd0324cd72519

a-palchikov commented on integration/integration_test.go:
```diff
@@ -1115,13 +1146,18 @@ func (s *IntSuite) runDisconnectTest(c *check.C, tc disconnectTestCase) {
 	select {
 	case <-time.After(tc.disconnectTimeout + time.Second):
-		c.Fatalf("timeout waiting for session to exit: %+v", tc)
+		pprof.Lookup("goroutine").WriteTo(os.Stderr, 1)
```
Usually, with the (longer) timeout that subsequently fails the test, it is helpful to know where the process is stuck, hence this is not trying to dump the current goroutine's stack but the processes state (e.g. akin to `go test` when it terminates the test process after a timeout).
So I would not expect `runtime.Stack(buf, true)` to be more appropriate here as there is no need to buffer the result - it is fine to stream it directly to stderr.

awly replied:
makes sense.
I assumed that using `pprof` here required special flags to `go test` to collect the profile, but it seems to work out-of-the-box.
this is much cleaner than `runtime.Stack` :+1:

a-palchikov committed:
Address review comments

r0mant commented on integration/app_integration_test.go:
```diff
@@ -238,6 +239,8 @@ func setup(t *testing.T) *pack {
 	tr := utils.NewTracer(utils.ThisFunction()).Start()
 	defer tr.Stop()
 
+	log := testlog.FailureOnly(t)
```
Does this mean integration tests are not going to output info/warn messages anymore, or is there a way to toggle them on? Just asking cause as I was working on an integration tests for db access, I used info outputs quite a bit to troubleshoot issues.

a-palchikov replied:
No, the test output will be streamed as expected in verbose mode.

r0mant commented on integration/integration_test.go:
```diff
@@ -119,7 +121,7 @@ func (s *IntSuite) SetUpSuite(c *check.C) {
 	// close & re-open stdin because 'go test' runs with os.stdin connected to /dev/null
 	stdin, err := os.Open("/dev/tty")
-	if err != nil {
+	if err == nil {
```
I'm curious if this should just assert `c.Assert(err, check.IsNil)`, or is it an expected scenario, do you know? Kinda weird that tests pass with both this condition and (presumably) erroneous previous condition.

a-palchikov replied:
This does look suspicious. The original implementation did just that and was changed [here](https://github.com/gravitational/teleport/commit/e4023519d0bea20e187e1666f933fe9ff72e8b7a).

Now, aside from that. the real question is why stdin is required in an integration test? As long as we can input the data from some file, it should be a good replacement for the code that'd require user input otherwise, so it's sort of confusing. I'll look at this in a later PR.

Regarding the logrus update - no, `WithError` will not wrap errors as you imagine since this is gravitational/trace functionality (namely, the TextFormatter implemented therein) which is only available in `trace` 1.1.10 and on. I'll bump it separately.

a-palchikov committed:
Move goroutine profile dump into a separate func to handle parameters consistently for all invocations
